### PR TITLE
Further Fido Refactoring

### LINKED ISSUE #2992: Fido.search errors very badly when no a.Time is specified
I am 90% sure this error is triggered because no `a.Time` is specified. If that is indeed the case the error that is raised is completely unhelpful.

```python
In [23]: Fido.search(a.Instrument("soon"))
```
```                                                                                                                                                                                           
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-23-e4b5b64d93c7> in <module>
----> 1 Fido.search(a.Instrument("soon"))

~/Git/sunpy/sunpy/net/fido_factory.py in search(self, *query)
    295         """  # noqa
    296         query = attr.and_(*query)
--> 297         return UnifiedResponse(query_walker.create(query, self))
    298 
    299     def fetch(self, *query_results, path=None, max_conn=5, progress=True,

~/Git/sunpy/sunpy/net/attr.py in create(self, *args, **kwargs)
    409 
    410     def create(self, *args, **kwargs):
--> 411         return self.createmm(self, *args, **kwargs)
    412 
    413     def apply(self, *args, **kwargs):

~/Git/sunpy/sunpy/util/multimethod.py in __call__(self, *args, **kwargs)
    119                 self.cache[types] = fun
    120                 return fun(*args, **kwargs)
--> 121         raise TypeError('{0!r}'.format(types))
    122 
    123     # XXX: Other Python implementations.

TypeError: (<class 'sunpy.net.vso.attrs.Instrument'>,)
```

### LINKED ISSUE #3368: Generalise QueryResponse
Currently each Fido client has a `QueryResponse` object.

* [VSO](https://github.com/sunpy/sunpy/blob/master/sunpy/net/vso/vso.py#L148)
* [JSOC](https://github.com/sunpy/sunpy/blob/master/sunpy/net/jsoc/jsoc.py#L34)
* [dataretriever](https://github.com/sunpy/sunpy/blob/master/sunpy/net/dataretriever/client.py#L55)

It should also be noted that the HEK client directly uses a customised version of `astropy.table`: https://github.com/sunpy/sunpy/blob/master/sunpy/net/hek/hek.py#L123

To address #3321 we need to allow each DR client to customise the columns in the `QueryResponse` object for their search, and return row metadata for each result. Providing an intermediate data structure to do this (i.e a list of dicts) seems just a complex, so it's probably easier to have the DR client return a `QueryReponse` object directly.

To facilitate this we would want the API of the `QueryReponse` object to be much cleaner than it is at the moment.

Given the use of `astropy.table` in all the `QueryReponse` objects for display and the customisation of astropy table in `HEK` I think something based closely on astropy table would be a good way of doing this.

### LINKED ISSUE #3814: Deprecation warning due to sunpy.net.vso.attrs
I ran test for `test_fido` file and this warning pops up:
```
sunpy/net/tests/test_fido.py:268: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <[AttributeError("'Time' object has no attribute 'start'",) raised in repr()] Time object at 0x7f7eff4ebf98>
args = ('2011-06-07 06:33', '2011-06-07 06:33:08'), kwargs = {}, name = 'Time'

    def __init__(self, *args, **kwargs):
        name = type(self).__name__
        warnings.warn(f"sunpy.net.vso.attrs.{name} is deprecated, please use sunpy.net.attrs.{name}",
>                     SunpyDeprecationWarning)
E       sunpy.util.exceptions.SunpyDeprecationWarning: sunpy.net.vso.attrs.Time is deprecated, please use sunpy.net.attrs.Time

sunpy/net/vso/attrs.py:380: SunpyDeprecationWarning
```
I think it can be simply solved by replacing `sunpy.net.vso.attrs` by `sunpy.net.attrs`

### LINKED ISSUE #3830: Allow Fido searches without time
I am currently trying to download some synoptic maps from JSOC. Because they is one map / Carrington rotation, they are indexed by Carrington number and *not* time. This means that the following search should be valid, but instead throws an error:

```python
from sunpy.net import Fido, attrs as a
import drms
c = drms.Client()  
print(c.pkeys('hmi.synoptic_mr_polfil_720s'))
res = Fido.search(a.jsoc.Series('hmi.synoptic_mr_polfil_720s'),
                  a.jsoc.PrimeKey('CAR_ROT', '2215'),
                  a.jsoc.Notify('sunpy@sunpy.org'))
```

The print statement prints `['CAR_ROT']`, I think confirming that time is not needed.

The error thrown is
```python
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-38-5774b7b06731> in <module>
      5 res = Fido.search(a.jsoc.Series('hmi.synoptic_mr_polfil_720s'),
      6                   a.jsoc.PrimeKey('CAR_ROT', '2215'),
----> 7                   a.jsoc.Notify('sunpy@sunpy.org'))  

~/github/sunpy/sunpy/net/fido_factory.py in search(self, *query)
    318         """  # noqa
    319         query = attr.and_(*query)
--> 320         return UnifiedResponse(query_walker.create(query, self))
    321 
    322     def fetch(self, *query_results, path=None, max_conn=5, progress=True,

~/github/sunpy/sunpy/net/attr.py in create(self, *args, **kwargs)
    457         Call the create function(s) matching the arguments to this method.
    458         """
--> 459         return self.createmm(self, *args, **kwargs)
    460 
    461     def apply(self, *args, **kwargs):

~/github/sunpy/sunpy/util/functools.py in wrapper(*args, **kwargs)
     17 
     18     def wrapper(*args, **kwargs):
---> 19         return dispatcher.dispatch(args[1].__class__)(*args, **kwargs)
     20 
     21     wrapper.dispatch = dispatcher.dispatch

~/github/sunpy/sunpy/net/fido_factory.py in _create_and(walker, query, factory)
    249         for at in query.attrs:
    250             error += str(at) + ', '
--> 251         raise ValueError(error)
    252 
    253     # Return the response and the client

ValueError: The following part of the query did not have a time specified:
<Series('hmi.synoptic_mr_polfil_720s')>, <PrimeKey('CAR_ROT','2215')>, <Notify('sunpy@sunpy.org')>, 
```

### PR DESCRIPTION
**Pull request by Cadair**
This PR fixes a number of things in net, I have gone to great lengths to keep the commits independent. If you want to review them one by one you should be able to. There are probably more tests that could do with writing, if you see some missing then let me know.

1) ~Allows clients to register new attrs namespaces, i.e. `a.dkist`~
1) Implements an ABC for `QueryResponse` and makes all the built in clients use it.
1) Allows Fido searches without Time, but makes sure all the clients that require Time don't respond to queries without it.
1) Allows multiple Fido clients to return results from a single AND section of a query. So both `RHESSIClient` and `VSO` now service queries for `a.Instrument("RHESSI")`.

fixes #3830 fixes #3368 fixes #2992 fixes #3814

### CODE REVIEW

**Comment by Cadair** [comment]
@nabobalis We need to work out how the attrs registration implemented here interacts with #3637

**Comment by nabobalis** [comment]
How do you want to work it out?

File: docs/guide/acquiring_data/fido.rst [hunk 1]
```diff
@@ -46,7 +46,6 @@ variable set to the Fido search, in this case, result::
     <BLANKLINE>
     3 Results from the LYRAClient:
          Start Time           End Time      Source Instrument Wavelength
-           str19               str19         str6     str4       str3
```
Comment (dstansby) [inline docs/guide/acquiring_data/fido.rst]: Can the change that gets rid of these bits be put in it's own PR?

File: sunpy/database/tests/test_tables.py [hunk 2]
```diff
@@ -204,7 +204,7 @@ def test_entries_from_fido_search_result_JSOC():
 @pytest.mark.remote_data
 def test_from_fido_search_result_block(fido_search_result):
     entry = DatabaseEntry._from_fido_search_result_block(
-        fido_search_result[0, 0][0].get_response(0)[0])
+        fido_search_result[0, 0][0].get_response(0).blocks[0])
```
Comment (nabobalis) [inline sunpy/database/tests/test_tables.py]: What caused this change and the `_data` change?

File: sunpy/net/_attrs.py [hunk 3]
```diff
@@ -227,3 +227,19 @@ class Resolution(SimpleAttr):
...
     pass
+
+
+class Physobs(SimpleAttr):
+    """
+    Specifies the physical observable the VSO can search for.
+
+    Parameters
+    ----------
+    value : str
```
Comment (nabobalis) [inline sunpy/net/_attrs.py]:
```suggestion
    value : `str`
```
Add a description?

**Comment by nabobalis** [comment]
> 1. Allows clients to register new attrs namespaces, i.e. `a.dkist`

Is this documented anywhere? 

> 3. Allows Fido searches without Time, but makes sure all the clients that require Time don't respond to queries without it.

Does each client have to manually account for this? 

> 4. Allows multiple Fido clients to return results from a single AND section of a query. So both `RHESSIClient` and `VSO` now service queries for `a.Instrument("RHESSI")`.

Will this not change our net examples?

File: docs/guide/acquiring_data/fido.rst (hunk 1 above)
Comment (nabobalis) [reply docs/guide/acquiring_data/fido.rst]: Any reason?

File: sunpy/net/dataretriever/sources/rhessi.py [hunk 4]
```diff
@@ -50,14 +50,13 @@ class RHESSIClient(GenericClient):
 
     >>> from sunpy.net import Fido, attrs as a
     >>> results = Fido.search(a.Time("2016/1/1", "2016/1/2"),
-    ...                       a.Instrument('RHESSI'))  #doctest: +REMOTE_DATA
+    ...                       a.Instrument("RHESSI"), a.Physobs("summary_lightcurve"))  #doctest: +REMOTE_DATA
```
Comment (nabobalis) [inline sunpy/net/dataretriever/sources/rhessi.py]: This count as a breaking change?

File: sunpy/net/dataretriever/sources/rhessi.py [hunk 5]
```diff
@@ -193,9 +192,23 @@ def _can_handle_query(cls, *query):
...
+        optional = {a.Physobs}
+
+        all_attrs = {type(x) for x in query}
+
+        query_attrs = set(type(x) for x in query)
+
+        if not required.issubset(query_attrs) or not all_attrs.issubset(query_attrs):
+            return False
+
+        matches = True
```
Comment (nabobalis) [inline sunpy/net/dataretriever/sources/rhessi.py]: Any reason to match true first?

File: sunpy/net/base_client.py [hunk 6]
```diff
@@ -1,7 +1,73 @@
...
+      I.e. it must always return an object of ``type(self)``.
+
+    """
+
+    @abstractmethod
+    def build_table(self):
+        """
+        Return an `astropy.table.Table` representation of the query response.
+        """
+
```
Comment (nabobalis) [inline sunpy/net/base_client.py]: Do we need to raise NotImplemented just to make sure they are in any subclass?

File: sunpy/database/tests/test_tables.py (hunk 2 above)
Comment (nabobalis) [reply sunpy/database/tests/test_tables.py]: I see it was the new ABC. 

Does this mean a user has to worry about this?

File: sunpy/net/fido_factory.py [hunk 7]
```diff
@@ -41,68 +40,28 @@ class UnifiedResponse(Sequence):
...
-                    self._numfile += len(block[0])
-                elif hasattr(block, 'client'):
-                    tmplst.append(block)
-                    self._numfile += len(block)
-                else:
-                    raise ValueError(
-                        "{} is not a valid input to UnifiedResponse.".format(type(lst)))
-        self._list = tmplst
+        for result in results:
+            if not isinstance(result, BaseQueryResponse):
```
Comment (nabobalis) [inline sunpy/net/fido_factory.py]: Would `issubclass` make more sense or does it change nothing?

File: sunpy/database/tests/test_tables.py (hunk 2 above)
Comment (Cadair) [reply sunpy/database/tests/test_tables.py]: It's the fact that slicing the `BaseQueryResponse` object should always return an instance of itself. In this current implementation I haven't standardised any underlying storage, so the way that each subclass of the base exposes a way to slice down to the underlying data is not specified. Perhaps I should add this to the class.

File: docs/guide/acquiring_data/fido.rst (hunk 1 above)
Comment (Cadair) [reply docs/guide/acquiring_data/fido.rst]: This change is tied up with standardising the `__str__`` and `__repr__` methods in `BaseQueryResponse`.

As for large diff, can I suggest you try and review the PR commit by commit if you want to have some logical separation between changes. I will add a changelog(s) to each commit in a bit to hopefully make it clearer as well.

File: sunpy/net/dataretriever/sources/rhessi.py (hunk 4 above)
Comment (Cadair) [reply sunpy/net/dataretriever/sources/rhessi.py]: yeah, doing `a.Time(), a.Instrument("RHESSI")` now returns results from both our `RHESSIClient` and `VSOClient` so it's giving you *more* than before. (If you *just* want `RHESSIClient` results you have to add this).

File: sunpy/net/dataretriever/sources/rhessi.py (hunk 5 above)
Comment (Cadair) [reply sunpy/net/dataretriever/sources/rhessi.py]: It's just the way I wrote the logic, could just as easily be the other way around :man_shrugging:

File: sunpy/net/fido_factory.py (hunk 7 above)
Comment (Cadair) [reply sunpy/net/fido_factory.py]: It should be passed instances (i.e. with results) rather than types here, so `isinstance` is really what we need.

**Comment by Cadair** [comment]
> > 1. Allows clients to register new attrs namespaces, i.e. `a.dkist`
> 
> Is this documented anywhere?

Not yet, I plan to fold it into #3768 if people think it's a good idea.

> 
> > 1. Allows Fido searches without Time, but makes sure all the clients that require Time don't respond to queries without it.
> 
> Does each client have to manually account for this?

Yes, clients which require a Time should specify this in their `_can_handle_query`. All the ones in SunPy do this now.

> 
> > 1. Allows multiple Fido clients to return results from a single AND section of a query. So both `RHESSIClient` and `VSO` now service queries for `a.Instrument("RHESSI")`.
> 
> Will this not change our net examples?

Yes, I think I have got them all though.

**Comment by nabobalis** [comment]
> Not yet, I plan to fold it into #3768 if people think it's a good idea.

Fine by me.

> Yes, clients which require a Time should specify this in their `_can_handle_query`. All the ones in SunPy do this now.

I assume one more thing to add to #3768?

> Yes, I think I have got them all though.

Did you do them in this PR? I did skip over them?

File: sunpy/net/dataretriever/sources/rhessi.py (hunk 5 above)
Comment (nabobalis) [reply sunpy/net/dataretriever/sources/rhessi.py]: Ignore me then.

File: sunpy/net/dataretriever/sources/rhessi.py (hunk 4 above)
Comment (nabobalis) [reply sunpy/net/dataretriever/sources/rhessi.py]: More changelogs!

File: sunpy/database/tests/test_tables.py (hunk 2 above)
Comment (nabobalis) [reply sunpy/database/tests/test_tables.py]: How much effort would that be?

**Comment by Cadair** [comment]
> I assume one more thing to add to #3768?

Yeah, not a bad plan.

> 
> > Yes, I think I have got them all though.
> 
> Did you do them in this PR? I did skip over them?

I thought I did, there weren't many.

**Comment by nabobalis** [comment]
> I thought I did, there weren't many.

Ah you mean doc examples, I was referring to our user guide.

**Comment by Cadair** [comment]
> Ah you mean doc examples, I was referring to our user guide.

Well there aren't any doctests failing... :grinning:

**Comment by nabobalis** [comment]
There is `/root/project/docs/code_ref/net.rst.rst:31:autosummary: stub file not found 'sunpy.net.attrs.Physobs'. Check your autosummary_generate setting.`

File: docs/guide/acquiring_data/fido.rst (hunk 1 above)
Comment (dstansby) [reply docs/guide/acquiring_data/fido.rst]: It looks like at least commits 2 and 3 could easily be simple individual PRs - I'd be happy to cherry-pick them out myself. Definitely appreciate that everything has to be here to start with to get the end product working, but I don't have the mental capacity to properly review a 500 line PR.

File: sunpy/database/tests/test_tables.py (hunk 2 above)
Comment (Cadair) [reply sunpy/database/tests/test_tables.py]: Not much, can be done.

Comment (Cadair) [reply sunpy/database/tests/test_tables.py]: It is done.

**Comment by Cadair** [comment]
Some info on the commit history here, we have the following changes:

1) [Split out VSO exceptions into their own file](https://github.com/sunpy/sunpy/pull/3770/commits/b3071f26c144e25a20f98e00dab6ba2a40d716f6) A super trivial change, just to see the wood from the trees in `vso.py`.
1) [Implement client specific attrs registration](https://github.com/sunpy/sunpy/pull/3770/commits/7ddbefbeb0ae07225d9604e967c967526a4bb112) This commit is independent of the others, and also is a small diff.
1) [Move physobs to be a top level attr](https://github.com/sunpy/sunpy/pull/3770/commits/7993a1fcca7919c2b6fe764843f7651a4827a1a5) This commit is also small, it is however required by the "Multiple client" commit as that commit requires `a.Physobs` for the `RHESSIClient`.
1) [Implement an ABC for QueryResponse](https://github.com/sunpy/sunpy/pull/3770/commits/ce87ecba402185818327c670a10a6918b8feef4e) this commit is a *large* diff. A large amount of the diff here is a concequence of me consolidating all the `__repr__` and `__str__` methods for `QueryResponse` objects into the base class.
1) [Fido Multiple client and no a.Time support](https://github.com/sunpy/sunpy/pull/3770/commits/946fa53243a39b4fa4dee4acf72b174d8455b296) This commit is also large, and it makes a couple of breaking changes to the behaviour of `RHESSIClient` and `EVEClient ` when used through `Fido`. This commit is dependant upon both the previous two because of the changes to the `RHESSIClient` and also changes to the `Fido` attr walker made in the QueryResponse commit.

The GitHub interface allows you to review each commit individually, and they now all come with changelog entries describing them in detail, this would be my preferred way for people to review this, but I can also split it into two PRs if needs be, however they will have interdependence on each other so aren't cleanly separable. (The first three commits here are separable with the last two, but the last two are 90% of the diff, so splitting it like that wouldn't shrink the diff significantly).

**Comment by dstansby** [comment]
> The GitHub interface allows you to review each commit individually

I just tried viewing a commit and approving, but it just approved the whole PR?

**Comment by Cadair** [comment]
> > The GitHub interface allows you to review each commit individually
> 
> I just tried viewing a commit and approving, but it just approved the whole PR?

That's true, but I meant you could filter down the diff and still comment and stuff.

File: sunpy/database/tables.py [hunk 8]
```diff
@@ -292,7 +292,7 @@ def _from_query_result_block(cls, qr_block, default_waveunit=None):
         >>> qr = client.search(
         ...     a.Time('2001/1/1', '2001/1/2'),
         ...     a.Instrument('eit'))  # doctest: +REMOTE_DATA
-        >>> entry = DatabaseEntry._from_query_result_block(qr[0])  # doctest: +REMOTE_DATA
+        >>> entry = DatabaseEntry._from_query_result_block(qr._data[0])  # doctest: +REMOTE_DATA
```
Comment (Cadair) [inline sunpy/database/tables.py]:
```suggestion
        >>> entry = DatabaseEntry._from_query_result_block(qr.blocks[0])  # doctest: +REMOTE_DATA
```

File: sunpy/net/base_client.py [hunk 9]
```diff
@@ -1,7 +1,82 @@
...
+            In general, this doesn't have to be the same instance of
+            ``BaseClient``, this is left to the client developer. If there is a
+            significant connection overhead in creating an instance of a client
+            you might want it to be the same instance as used for the search.
+        """
+
+    @client.setter
+    @abstractmethod
+    def client(self):
```
Comment (aribalam) [inline sunpy/net/base_client.py]: Should we not add an argument for passing in an client?

**Comment by Cadair** [comment]
>     * #3830 (a search without time)

:+1: 

>     * #2992 (a nice error message thrown when `a.Time` not provided)

This fixes this issue by never raising an error in this situation.

>     * #3733 (that post search filtering works for Fido - I couldn't see  a test in `test_fido.py`)

I don't know why I said this fixes that issue, it really doesn't I have removed it from the top comment.

**Comment by Cadair** [comment]
@dstansby I added a test for the no Time search.

**Comment by Cadair** [comment]
(Current test fails look like flaky servers, wont re-run to save electrons)

File: changelog/3770.breaking.1.rst [hunk 10]
```diff
@@ -0,0 +1 @@
+``Fido.search`` will now return results from all clients which match a query, you no longer have to make the query specific to a single client. This means that searches involving the 'eve' and 'rhessi' instruments will now potentially also return results from the VSO. For `~sunpy.net.dataretriever.RHESSIClient` you can now specify ``a.Physobs("summary_lightcurve")`` to exclude VSO results.
```
Comment (hayesla) [inline changelog/3770.breaking.1.rst]: `instruments will now potentially` is the potentially due to the fact that it will return VSO results as well as the client results if they exist? maybe wording should be more along the lines of, `the search will also return results from vso if they exist`?

**Comment by Cadair** [comment]
I have popped #3877 out into it's own PR. I think the last two remaining (after that) changes are maybe best reviewed together as they both touch the same Fido attr walker part of the code.

**Comment by dstansby** [comment]
Great, I think I just merged the last independent one? I have a train later this afternoon, so should have a decent chunk of time to review this 👍

File: sunpy/net/fido_factory.py [hunk 11]
```diff
@@ -111,7 +70,9 @@ def __getitem__(self, aslice):
         returned from those clients.
         """
         # Just a single int as a slice, we are just indexing client.
-        if isinstance(aslice, (int, slice)):
+        if isinstance(aslice, int):
+            ret = [self._list[aslice]]
```
Comment (dstansby) [inline sunpy/net/fido_factory.py]: Could instead do `aslice = slice(aslice, aslice + 1)` here, then remove the line below?

File: sunpy/net/fido_factory.py [hunk 12]
```diff
@@ -317,7 +265,17 @@ def search(self, *query):
         parts individually.
         """  # noqa
         query = attr.and_(*query)
-        return UnifiedResponse(query_walker.create(query, self))
+        results = query_walker.create(query, self)
+
+        # If we have searched the VSO but no results were returned, but another
+        # client generated results, we drop the empty VSO results for tidiness.
```
Comment (dstansby) [inline sunpy/net/fido_factory.py]: This behaviour should be documented somewhere in the user facing docs

I also wondering if it should be generalised - if at least one client returns results, should all other zero-result clients be dropped?

Comment (Cadair) [reply sunpy/net/fido_factory.py]: VSO is special cased in a way, it's `_can_handle_query` is exceptionally broad (mostly for historical reasons). This therefore is a way of bringing VSO into line with the other clients which afaik don't have their `_can_handle_query` return True for a query they will not return results for.

File: sunpy/net/dataretriever/sources/eve.py [hunk 13]
```diff
@@ -100,14 +99,36 @@ def _can_handle_query(cls, *query):
...
             answer as to whether client can service the query
         """
-        chk_var = 0
-        for x in query:
-            if x.__class__.__name__ == 'Instrument' and x.value.lower() == 'eve':
-                chk_var += 1
+        from sunpy.net import attrs as a
+
+        required = {a.Time, a.Instrument, a.Level}
+        optional = {}  # Level should really be in here
```
[the remark this answers is not in the document]
Comment (Cadair) [reply sunpy/net/dataretriever/sources/eve.py]: :+1: (It's because VSO doesn't handle EVE level correctly (yet))

File: sunpy/database/tables.py [hunk 14]
```diff
@@ -292,7 +292,7 @@ def _from_query_result_block(cls, qr_block, default_waveunit=None):
         >>> qr = client.search(
         ...     a.Time('2001/1/1', '2001/1/2'),
         ...     a.Instrument('eit'))  # doctest: +REMOTE_DATA
-        >>> entry = DatabaseEntry._from_query_result_block(qr[0])  # doctest: +REMOTE_DATA
```
[the remark this answers is not in the document]
Comment (Cadair) [reply sunpy/database/tables.py]: FTR this is a docstring of a private method :rofl:

File: sunpy/net/base_client.py [hunk 15]
```diff
@@ -1,8 +1,82 @@
...
+
+    @property
+    @abstractmethod
+    def client(self):
+        """
+        An instance of BaseClient used to generate the results.
+
+        Generally this is used to fetch the results later.
+
+        .. note::
```
[the remark this answers is not in the document]
Comment (Cadair) [reply sunpy/net/base_client.py]: I opted for this because I think the more obvious box is useful as opposed to another heading, but :man_shrugging:

File: sunpy/net/dataretriever/tests/test_eve.py [hunk 16]
```diff
@@ -45,6 +45,15 @@ def test_can_handle_query():
...
         Time('2012/8/9', '2012/8/10'), Instrument('eve'), Source('sdo'))
     assert ans3 is False
+    ans4 = eve.EVEClient._can_handle_query(
+        Time('2012/8/9', '2012/8/10'), Instrument('eve'), Level('0CS'))
+    assert ans4 is True
+    ans5 = eve.EVEClient._can_handle_query(
+        Time('2012/8/9', '2012/8/10'), Instrument('eve'), Level('wibble'))
+    assert ans5 is False
+    ans6 = eve.EVEClient._can_handle_query(
+        Time('2012/8/9', '2012/8/10'), Instrument('eve'), Level(0.5))
```
[the remark this answers is not in the document]
Comment (Cadair) [reply sunpy/net/dataretriever/tests/test_eve.py]: Maybe, or it's that no clients we have support it. Either way, the test is really here to check that floats work as expected :grin:

File: sunpy/net/fido_factory.py (hunk 12 above)
Comment (Cadair) [reply sunpy/net/fido_factory.py]: I have updated the comment to elaborate on why.

File: sunpy/net/dataretriever/client.py [hunk 17]
```diff
@@ -53,19 +53,46 @@ def iter_urls(amap, url_list, time):
...
+            item = slice(item, item+1)
+        return type(self)(self._data[item], client=self.client)
+
+    def __iter__(self):
+        for block in self._data:
+            yield block
 
     @classmethod
-    def create(cls, amap, lst, time=None):
+    def create(cls, amap, lst, time=None, *, client):
```
[the remark this answers is not in the document]
Comment (Cadair) [reply sunpy/net/dataretriever/client.py]: I have now changed it to `client=None`

File: changelog/3770.breaking.1.rst (hunk 10 above)
[the remark this answers is not in the document]
Comment (Cadair) [reply changelog/3770.breaking.1.rst]: I have updated this wording to hopefully be a little less confusing. I am generally against any way of directly specifying a client. The reasoning behind this is that the user should be requesting the physical parameters of the data they want, not caring about what client is servicing the request.[1] The change here (highlighted hopefully by the new wording) is that we are now providing this `a.Physobs` to allow you to just select this one data product (which happens to not be in the VSO).

If a future X ray mission provided a similar data product one could re-use this and then get all the summary lightcurves you want. :grin:

**Comment by Cadair** [comment]
Where's codecov.

**Comment by Cadair** [comment]
:rocket: Thanks everyone!
